contrib/hydra/NpinsInput.pm

```diff
@@ -0,0 +1,215 @@
+use Fcntl qw(:flock);
+use Digest::SHA qw(sha256_hex);
+use File::Slurper qw(read_text);
+use JSON::MaybeXS qw(decode_json);
+use Data::Dumper;
+
+## no critic (InputOutput::RequireBriefOpen)
+## no critic (CodeLayout::ProhibitParensWithBuiltins)
+## no critic (ControlStructures::ProhibitCascadingIfElse)
+## no critic (NamingConventions::Capitalization)
+## no critic (Subroutines::ProhibitExcessComplexity, Subroutines::ProhibitManyArgs)
+## no critic (ValuesAndExpressions::ProhibitConstantPragma)
+
+our $VERSION = 1;
+use constant DEFAULT_GIT_TIMEOUT => 600;
+
+sub _print_if_debug {
+    my ($evaluation, $msg) = @_;
+
+    if (not $ENV{'HYDRA_DEBUG'}) {
+        return;
+    }
+
+    my $evalid = $evaluation->id;
+    my $project = $evaluation->jobset->project->name;
+    my $jobset = $evaluation->jobset->name;
+
+    print {*STDERR} "NpinsInput: [Eval $evalid of $project:$jobset] $msg\n" || croak;
+    return;
+}
+
+sub _print {
+    my ($evaluation, $msg) = @_;
+    my $evalid = $evaluation->id;
+    my $project = $evaluation->jobset->project->name;
+    my $jobset = $evaluation->jobset->name;
+
+    print {*STDERR} "NpinsInput: [Eval $evalid of $project:$jobset] $msg\n" || croak;
+    return;
+}
+
+sub supportedInputTypes {
+    my ($self, $input_types) = @_;
+    $input_types->{'npins-channel'} = 'Npins channel';
+    $input_types->{'npins-pypi'} = 'Npins PyPi';
+    return;
+}
+
+sub evalAdded {
+    my ($self, $trace_id, $jobset, $evaluation) = @_;
+
+    # Find pins directory we need to act on
+    my $npins_dir_input = $evaluation->jobsetevalinputs->find({ name => 'npins-directory', type => 'string' });
+    my $npins_dir_name = 'npins';
+    if (defined($npins_dir_input)) {
+        $npins_dir_name = $npins_dir_input->get_column('value');
+    }
+
+    # Find GitHub host
+    my $github_host_input = $evaluation->jobsetevalinputs->find({ name => 'npins-github-host', type => 'string' });
+    my $github_host = 'https://github.com';
+    if (defined($github_host_input)) {
+        $github_host = $github_host_input->get_column('value');
+    }
+
+    # Find the primary input and its sources.json
+    my $repo_root = $evaluation->jobsetevalinputs->find({ name => $evaluation->nixexprinput, altnr => 0 });
+    if (not defined($repo_root)) {
+        _print_if_debug($evaluation, 'Skipping as there does not seem to be a main input');
+        return;
+    }
+    $repo_root = $repo_root->path;
+    my $npins_dir = "$repo_root/$npins_dir_name";
+    my $sources_json = "$npins_dir/sources.json";
+    _print_if_debug($evaluation, "Will look for sources in $sources_json");
+    if (not -f $sources_json) {
+        _print_if_debug($evaluation, "No sources.json in $npins_dir_name/");
+        return;
+    }
+
+    # Parse JSON
+    my $json_contents;
+    if (not eval { $json_contents = decode_json(read_text($sources_json)); 1 }) {
+        _print($evaluation, 'Invalid JSON');
+        return;
+    }
+
+    if ($json_contents->{version} != 2) {
```

Comment:
Unfortunately, this isn't compatible with the current version of npins: https://github.com/andir/npins/commit/fe8e6fe06c17529de7dcf67c71ca43c7adc6a50c ;)